### PR TITLE
Support CORS GET Requests

### PR DESCRIPTION
**Pull request by jmcanterafonseca**
This PR is intended to enable cross-origin GET requests from Web Browsers to Orion. The advantage of supporting CORS is that Orion will be available directly from a Web Browser client without the need of intermmediate backends or proxies. More info on CORS can be found at

https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS
https://annevankesteren.nl/2015/02/same-origin-policy

For the time being, only the ability to perform GET requests have been implemented. For POST / PUT requests it would be needed to implement preflight requests through the OPTIONS method. But that would part of a subsequent PR. 

I guess the PR would need some unit tests as well, but I am a bit lost on how to start with. If they are needed any hint on how to do that would be appreciated.

### CODE REVIEW

**Comment by fgalan** [comment]
Thanks for the contribution! This PR could be an effective solution to issue https://github.com/telefonicaid/fiware-orion/issues/501 and I'm sure many ussers will be happy with that ;)

I would need some time to read the references you cite to learn more on CORS and also to discuss the PR with @kzangeli . Regarding testing for this PR, rather than unit test I'd suggest to use test harness (a kind of end-to-end automatic test framework we have for Orion), but don't worry about that too much: @kzangeli or me could manage that.

File: src/lib/rest/restReply.cpp [hunk 1]
```diff
@@ -85,6 +87,15 @@ void restReply(ConnectionInfo* ciP, const std::string& answer)
       MHD_add_response_header(response, "Content-Type", "application/xml");
     else if (ciP->outFormat == JSON)
       MHD_add_response_header(response, "Content-Type", "application/json");
+
+    // If any origin is allowed the header is sent always with "any" as value
+    if(strcmp(allowedOrigin, "*") == 0) {
```
Comment (fgalan) [inline src/lib/rest/restReply.cpp]: Please use this C++ style:

```
if (...)
{
...
}
else
{
...
}
```

**Comment by jmcanterafonseca** [comment]
thanks for the feedback. I have updated the PR to meet the indenting rules and I have improved the design of the condition check.

**Comment by jmrobles** [comment]
+1! :)

**Comment by nifkuza** [comment]
+2

**Comment by fgalan** [comment]
The PR is now ready for review and final merge. LGTM from @kzangeli and @jmcanterafonseca are required.

File: src/app/contextBroker/contextBroker.cpp [hunk 2]
```diff
@@ -246,6 +247,7 @@ int             writeConcern;
 #define HTTPSCERTFILE_DESC  "certificate key file (for https)"
 #define RUSH_DESC           "rush host (IP:port)"
 #define MULTISERVICE_DESC   "service multi tenancy mode"
+#define ALLOWED_ORIGIN_DESC "CORS allowed origin. use '__ALL' for any"
```
Comment (fgalan) [inline src/app/contextBroker/contextBroker.cpp]: Allowing `-corsOrigin '*'` in the CLI is dangerour... If the user forget to include `*` as literal, the shell will extand it with the contents of the current directory. @jmcanterafonseca do you see any problem in using an alias ("__ALL" could be changed for a better one if you want)? Of course, this precludes to use literarly __ALL as domain origin is a name weird enough so nobody should be using it as real domain name :)

File: test/functionalTest/cases/501_cors/get_any_origin.test [hunk 3]
```diff
@@ -0,0 +1,134 @@
+# Copyright 2014 Telefonica Investigacion y Desarrollo, S.A.U
```
Comment (fgalan) [inline test/functionalTest/cases/501_cors/get_any_origin.test]: @jmcanterafonseca: have specially a look to the .test files included in this PR in order to check if the CB, after being started with SHELL-INIT and feed with what is shown in SHELL section should return the output shown in REGEXPECTED. 

I think the syntax of the test specification is quite straighforward, but if you have doubts, don't hesitate to contact me offline.

File: test/functionalTest/cases/501_cors/get_any_origin.test [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+orionCurl --url /v1/queryContext --payload "${payload}" --json
+echo
+echo
+
+--REGEXPECT--
+01. Update/APPEND E1-T1-A1
+==========================
+HTTP/1.1 200 OK
+Content-Length: 383
+Access-Control-Allow-Origin: *
```
Comment (fgalan) [inline test/functionalTest/cases/501_cors/get_any_origin.test]: @jmcanterafonseca: This is the response to a POST operation and includes the `Access-Control-Allow-Origin`. Taking into account the PR description, I thought that we are implemented CORS only for GETs... Is this correct?

File: src/lib/rest/rest.cpp [hunk 5]
```diff
@@ -1068,6 +1071,8 @@ void restInit
   rushHost      = _rushHost;
   rushPort      = _rushPort;
 
+  strcpy(restAllowedOrigin, _allowedOrigin);
```
Comment (kzangeli) [inline src/lib/rest/rest.cpp]: Use strncpy instead. The style guide script will complain ...
Wish we had the style guide check running ...

Comment (fgalan) [inline src/lib/rest/rest.cpp]: Fixed in f54a492

File: test/functionalTest/cases/501_cors/get_any_origin.test (hunk 4 above)
Comment (kzangeli) [inline test/functionalTest/cases/501_cors/get_any_origin.test]: That's pretty much what was said in the meeting in Madrid, right? Only GET (for now)

**Comment by kzangeli** [comment]
LGTM

File: test/functionalTest/cases/501_cors/get_any_origin.test (hunk 4 above)
Comment (jmcanterafonseca) [inline test/functionalTest/cases/501_cors/get_any_origin.test]: yup

File: test/functionalTest/cases/501_cors/get_any_origin.test (hunk 3 above)
Comment (jmcanterafonseca) [inline test/functionalTest/cases/501_cors/get_any_origin.test]: I see, thanks for the hint

File: src/app/contextBroker/contextBroker.cpp (hunk 2 above)
Comment (jmcanterafonseca) [inline src/app/contextBroker/contextBroker.cpp]: yes, you are right. I don't have any particular opinion on what syntax to use for representing any URI, if you are ok with __ALL , I'm fine

thanks!

**Comment by jmcanterafonseca** [comment]
LGTM as well, thanks for the work!

File: test/functionalTest/cases/501_cors/get_any_origin.test (hunk 4 above)
Comment (fgalan) [inline test/functionalTest/cases/501_cors/get_any_origin.test]: So you mean that `Access-Control-Allow-Origin` shouldn't be included in this case? In that case, I understand that the new code added to restReply.cpp should check that the request method was GET and only add the header in that case (I think you can get the method from the `ciP` object, maybe @kzangeli could provide more info on this).

File: src/app/contextBroker/contextBroker.cpp (hunk 2 above)
Comment (fgalan) [inline src/app/contextBroker/contextBroker.cpp]: Ok, let's keep "__ALL".

NTC

File: test/functionalTest/cases/501_cors/get_any_origin.test (hunk 4 above)
Comment (kzangeli) [inline test/functionalTest/cases/501_cors/get_any_origin.test]: ciP contains the method, yes

Comment (fgalan) [inline test/functionalTest/cases/501_cors/get_any_origin.test]: Fixed in 46f6cee4d627c4310fb6a470c7917c801023df2f

**Comment by fgalan** [comment]
LGTM (final)

**Pull request merged**